azure-iot-device/tests/provisioning/internal/test_request_response_provider.py

```diff
@@ -0,0 +1,151 @@
+    assert mock_mqtt_state_based_provider.publish.call_args[1][
+        "topic"
+    ] == fake_request_topic.format(fake_rid)
+    assert mock_mqtt_state_based_provider.publish.call_args[1]["message"] == req
+
+
+def test_subscribe_calls_subscribe_on_state_based_provider_with_provided_callback(
+    request_response_provider
+):
+    mock_mqtt_state_based_provider = request_response_provider._state_based_provider
+    mock_callback = MagicMock()
+    request_response_provider.subscribe(fake_subscribe_topic, mock_callback)
+    mock_mqtt_state_based_provider.subscribe.assert_called_once_with(
+        topic=fake_subscribe_topic, callback=mock_callback
+    )
+
+
+def test_subscribe_calls_subscribe_on_state_based_provider(request_response_provider):
+    mock_mqtt_state_based_provider = request_response_provider._state_based_provider
+    request_response_provider.subscribe(fake_subscribe_topic)
+    mock_mqtt_state_based_provider.subscribe.assert_called_once_with(
+        topic=fake_subscribe_topic, callback=request_response_provider._on_subscribe_completed
+    )
+
+
+def test_unsubscribe_calls_unsubscribe_on_state_based_provider_with_provided_callback(
+    request_response_provider
+):
+    mock_mqtt_state_based_provider = request_response_provider._state_based_provider
+    mock_callback = MagicMock()
+    request_response_provider.unsubscribe(fake_subscribe_topic, mock_callback)
+    mock_mqtt_state_based_provider.unsubscribe.assert_called_once_with(
+        topic=fake_subscribe_topic, callback=mock_callback
+    )
+
+
+def test_unsubscribe_calls_unsubscribe_on_state_based_provider(request_response_provider):
+    mock_mqtt_state_based_provider = request_response_provider._state_based_provider
+    request_response_provider.unsubscribe(fake_subscribe_topic)
+    mock_mqtt_state_based_provider.unsubscribe.assert_called_once_with(
+        topic=fake_subscribe_topic, callback=request_response_provider._on_unsubscribe_completed
+    )
+
+
+def test_on_provider_message_received_receives_response_and_calls_callback(
+    request_response_provider
+):
+    mock_mqtt_state_based_provider = request_response_provider._state_based_provider
+    req = "Leviosa"
+
+    mock_callback = MagicMock()
+    request_response_provider.send_request(
+        rid=fake_rid, topic=fake_request_topic.format(fake_rid), request=req, callback=mock_callback
+    )
+    assigning_status = "assigning"
+
+    payload = '{"operationId":"' + fake_operation_id + '","status":"' + assigning_status + '"}'
+
+    mock_payload = payload.encode("utf-8")
+    mock_mqtt_state_based_provider.on_state_based_provider_message_received(
+        fake_success_response_topic, mock_payload
+    )
+
+    topic_parts = fake_success_response_topic.split("$")
+    key_value_dict = urllib.parse.parse_qs(topic_parts[POS_QUERY_PARAM_PORTION])
+    mock_callback.assert_called_once_with(topic_parts[POS_URL_PORTION], key_value_dict, payload)
```

Comment:
>assert_called_once_with [](start = 18, length = 23)

Here and everywhere, please do NOT use function. Ever. It's very dangerous.
[https://engineeringblog.yelp.com/2015/02/assert_called_once-threat-or-menace.html](https://engineeringblog.yelp.com/2015/02/assert_called_once-threat-or-menace.html)

I've neglected to capture this in our wiki as a documented practice, that's my bad. I'll make sure to do so when we further formalize testing. In the interim, please remove this everywhere it is.

Reply:
I think I read this article long back and I know what you are saying. I think the issue was with assert_called_once and not assert_called_once_with.  The assert_called_once does not test anything

However if you have an assert_called_once_with(args) and assrt_Called_with(args) it actually works...

I have tested this on a test before.  I can still change this but I actually like the with ones 

---
In reply to: [282987319](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r282987319) [](ancestors = 282987319)

Reply:
No, the worst problem is with "assert_called_once_with" for various reasons including the ones you point out, but there's a fundamental underlying problem with all mock "assert_<x>" methods - if you make a typo in the method name it 1) still runs as if everything is fine and 2) causes the test to always pass.

i.e. if you make any typo your test will appear to work, but actually is not testing anything, and could be thus hiding bugs in the codebase.

Thus it is __much__ safer to use the build in `assert` command

---
In reply to: [283072222](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r283072222) [](ancestors = 283072222,282987319)

Reply:
I understand but all these things are true for `assert called once` and not with `assert called once with` or any of the other asserts as far as I know.
The spelling mistake has already happened to me with `assert called once`. That is when I discovered the page that you are referring to.
I made a conscious effort after that not to use `assert called once`
If I ever discover that I am mistaken I will change it...
I actually like `assert_called_once_with` as I can directly pass the arguments. 

---
In reply to: [283118252](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r283118252) [](ancestors = 283118252,283072222,282987319)

Reply:
The issue isn't specific to any given method, it's a fundamental problem of MagicMock().

However, I've checked out the source code for mock. It looks like it's been updated in recent years to explicitly disregard anything unknown that starts with `assert`, or, most importantly, the typo `assret` in the last couple years. So `assret_called_once_with` no longer will pass like it used to, nor does `assert_called_ocne_with` or other similar variations.

Technically, we could still hit the issue, with, say `asesrt_called_once_with`, since it begins with neither `assert` nor `assret` (the only two spellings explicitly guarded against), but I feel like that's a less likely to slip through. If this is a risk everyone else is okay with (see if Bert and Pierre are on board), I would be okay with reintroducing the use of the mock assertion methods.

However, the other note would be that assertion methods do fundamentally go against pytest syntactical patterns. Not a big deal, but using a method to assert rather than the `assert` statement is very unittest as opposed to pytest. But multiple statements to check the same thing isn't great either, so, perhaps that's a wash.

Like I said, I'm good with this if everyone else is.

---
In reply to: [283536174](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r283536174) [](ancestors = 283536174,283118252,283072222,282987319)

Reply:
will discuss and address them if needed in a later PR

---
In reply to: [283823577](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r283823577) [](ancestors = 283823577,283536174,283118252,283072222,282987319)